### PR TITLE
Release v0.4.0

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Release summary

Version number: v0.4.0

## Release details

This release:
- Updates message handling in skills
- Replaces serializer implementation; all serialization is now performed framework side
- Updates all skills for compatibility with new message handling
- Updates all protocols and protocol generator
- Updates package loading mechnanism
- Adds p2p_libp2p_client connection
- Fixes CLI bugs and refactors CLI
- Adds eject command to CLI
- Exposes identity and connection cryptos to all connections
- Updates connection loading mechanism
- Updates all connections for compatibility with new loading mechanism
- Extracts multiplexer into its own module
- Implements list all CLI command 
- Updates wallet to split into several crypto stores
- Refactors component registry and resources
- Extends soef connection functionality
- Implements AEABuilder reentrancy
- Updates p2p_libp2p connection
- Adds support for configurable runtime
- Refactors documentation
- Multiple docs updates
- Multiple test stability fixes

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `master` branch (left side), from `develop`
- [x] Lint and unit tests pass locally
- [x] I have checked the fingerprint hashes are correct by running (`scripts/generate_ipfs_hashes.py`)
- [x] I have regenerated the latest API docs
- [x] I built the documentation and updated it with the latest changes
- [x] I have added an item in `HISTORY.md` for this release
- [x] I bumped the version number in the `aea/__version__.py` file.
- [x] I bumped the version number in the `docs/version.md` file
- [x] I bumped the version number in every Docker image of the repo and published it. Also, I built and published them with tag `latest`  
      (check the READMEs of [`aea-develop`](../develop-image/README.md#publish) 
      and [`aea-deploy`](../deploy-image/README.md#publish))
- [x] I have checked that the documentation about the `aea cli` tool works
- [ ] I have pushed the latest packages to the registry.

## Further comments

-

### CODE REVIEW

File: aea/aea_builder.py [hunk 1]
```diff
@@ -1196,30 +1227,26 @@ def _load_and_add_skills(self, context: AgentContext) -> None:
...
+                    Skill,
+                    load_component_from_config(
+                        ComponentType.SKILL, configuration, skill_context=skill_context
+                    ),
+                )
+            resources.add_component(skill)
 
     def _load_connection(
-        self, address: Address, configuration: ConnectionConfig
+        self, identity: Identity, wallet: Wallet, configuration: ConnectionConfig
```
Comment (solarw) [inline aea/aea_builder.py:1243]: every connection will have access to the wallet?

File: aea/aea_builder.py [hunk 2]
```diff
@@ -1228,46 +1255,34 @@ def _load_connection(
...
-                address=address, configuration=configuration
+        else:
+            configuration = deepcopy(configuration)
+            return cast(
+                Connection,
+                load_component_from_config(
+                    ComponentType.CONNECTION,
+                    configuration,
+                    identity=identity,
+                    crypto_store=wallet.connection_cryptos,
```
Comment (solarw) [inline aea/aea_builder.py:1273]: what about to filter connection cryptos by connection id?

File: aea/aea_builder.py (hunk 1 above)
Comment (DavidMinarsch) [reply aea/aea_builder.py:1243]: No; only the Connection `CryptoStore`

File: aea/aea_builder.py (hunk 2 above)
Comment (DavidMinarsch) [reply aea/aea_builder.py:1273]: Perhaps in the future yes; as part of this release it won't happen.

**Pull request merged**